src/fw_update.c

```diff
@@ -164,6 +168,43 @@ static void on_ota_manifest(struct golioth_client *client,
     golioth_sys_sem_give(_manifest_rcvd);
 }
 
+static void backoff_reset(struct fw_update_component_context *ctx)
+{
+    ctx->backoff_duration_ms = 0;
+    ctx->last_fail_ts = 0;
+}
+
+static void backoff_increment(struct fw_update_component_context *ctx)
+{
+    if (ctx->backoff_duration_ms == 0)
+    {
+        ctx->backoff_duration_ms = BACKOFF_DURATION_INITIAL_MS;
+    }
+    else
+    {
+        ctx->backoff_duration_ms = ctx->backoff_duration_ms * 2;
+
+        if (ctx->backoff_duration_ms > BACKOFF_DURATION_MAX_MS)
+        {
+            ctx->backoff_duration_ms = BACKOFF_DURATION_MAX_MS;
+        }
+    }
+
+    ctx->last_fail_ts = golioth_sys_now_ms();
+}
+
+static int32_t backoff_ms_before_expiration(struct fw_update_component_context *ctx)
+{
+    uint64_t actual_duration = golioth_sys_now_ms() - ctx->last_fail_ts;
```

Comment:
I wrote this with rollover in mind. Though I did not realize we don't have access to the full bit-width of the type being return.  Since I'm returning an int32 for the duration, why don't I just mask the uint64 tick count to uint32 before calculation? I think it will perform the same way.